src/ray/core_worker/test/memory_store_test.cc

```diff
@@ -217,6 +217,124 @@ TEST(TestMemoryStore, TestObjectAllocator) {
+  auto plasma_store_object = ray::RayObject{rpc::ErrorType::OBJECT_IN_PLASMA};
+
+  // Object 0 is ready in memory store
+  // Object 1 is in plasma
+  // Object 2 is in plasma
+  // Object 3 is ready in memory store
+  // num_objects is 2 (expect 1 and 4 in ready, 1 and 2 in plasma_object_ids)
+  std::vector<ObjectID> object_ids = {ObjectID::FromRandom(),
+                                      ObjectID::FromRandom(),
+                                      ObjectID::FromRandom(),
+                                      ObjectID::FromRandom()};
+  absl::flat_hash_set<ObjectID> object_ids_set = {object_ids.begin(), object_ids.end()};
+  int num_objects = 2;
+
+  memory_store->Put(memory_store_object, object_ids[0]);
+  memory_store->Put(plasma_store_object, object_ids[1]);
+  memory_store->Put(plasma_store_object, object_ids[2]);
+  memory_store->Put(memory_store_object, object_ids[3]);
+
+  // DO WAIT
+  const auto status_or_results =
+      memory_store->Wait(object_ids_set, num_objects, 1000, ctx);
+  const auto &[ready, plasma_object_ids] = status_or_results.value();
+
+  // ASSERTIONS
+  ASSERT_TRUE(status_or_results.ok());
+  ASSERT_EQ(ready.size(), 2);
+  ASSERT_TRUE(ready.contains(object_ids[0]) && ready.contains(object_ids[3]));
+  ASSERT_EQ(plasma_object_ids.size(), 2);
+  ASSERT_TRUE(plasma_object_ids.contains(object_ids[1]) &&
+              plasma_object_ids.contains(object_ids[2]));
+}
+
+TEST(TestMemoryStore, TestWaitWithWaiting) {
+  // SETUP
+  InstrumentedIOContextWithThread io_context("TestWait");
+  std::shared_ptr<CoreWorkerMemoryStore> memory_store =
+      std::make_shared<CoreWorkerMemoryStore>(io_context.GetIoService());
+  WorkerContext ctx =
+      WorkerContext(WorkerType::WORKER, WorkerID::FromRandom(), JobID::FromInt(1));
+  const std::string buffer = "hello";
+  std::vector<rpc::ObjectReference> nested_refs;
+  auto memory_store_object =
+      ray::RayObject{MakeLocalMemoryBufferFromString(buffer), nullptr, nested_refs, true};
+  auto plasma_store_object = ray::RayObject{rpc::ErrorType::OBJECT_IN_PLASMA};
+
+  // Object 0 is ready in memory store
+  // Object 1 is in plasma
+  // Object 2 will be in plasma after 500ms
+  // Object 3 will be in ready in memory store after 500ms
+  std::vector<ObjectID> object_ids = {ObjectID::FromRandom(),
+                                      ObjectID::FromRandom(),
+                                      ObjectID::FromRandom(),
+                                      ObjectID::FromRandom()};
+  absl::flat_hash_set<ObjectID> object_ids_set = {object_ids.begin(), object_ids.end()};
+  int num_objects = 4;
+
+  memory_store->Put(memory_store_object, object_ids[0]);
+  memory_store->Put(plasma_store_object, object_ids[1]);
+
+  // DO WAIT AND ADD OBJECT 2 and 3
+  auto future = std::async(
+      std::launch::async,
+      [&memory_store, &object_ids, &plasma_store_object, &memory_store_object]() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(500));
```

Comment:
500ms is too slow for a unit test (we should be able to run hundreds of unit tests in <1s)

I'd suggest to reverse the pattern here and run the `Wait` call in the future. Then you can do some more fine grained assertions, for example:
```
auto future = <launch wait>;
ASSERT_EQ(future.wait_for(1ms), std::future_status::timeout);
memory_store->Put(plasma_store_object, object_ids[2]);
ASSERT_EQ(future.wait_for(1ms), std::future_status::timeout);
memory_store->Put(plasma_store_object, object_ids[3]);
future.get();
```